Add Glyph components public methods

chrissimpkins opened the pull request:
Adds a few QoL methods around Glyph components.

cmyr commented on src/glyph/mod.rs:
```diff
@@ -104,6 +104,28 @@ impl Glyph {
+    }
+
+    /// Returns a usize that represents the number of [`Component`] defined on the Glyph.
+    pub fn component_number(&self) -> usize {
```
I have a minor preference for:

```suggestion
    pub fn component_count(&self) -> usize {
```

chrissimpkins replied:
Ty!

cmyr commented on src/glyph/mod.rs:
```diff
@@ -104,6 +104,28 @@ impl Glyph {
+    }
+
+    /// Return a Vec of immutable [`Component`] references filtered by base glyph name.
+    pub fn get_components_with_base(&self, basename: &str) -> Vec<&Component> {
+        self.components.iter().filter(|x| *x.base == *basename).collect::<Vec<&Component>>()
+    }
```
Instead of returning a `Vec` (and allocating) I would prefer to return an `Iterator` and then give the caller the choice of allocating if necessary.

This ends up being slightly complicated for subtle lifetime reasons: we're returning an iterator that itself returns references to items owned by self, but the iterator *itself* also needs to outlive `basename`, since we need to compare with `basename` each time we `filter`.

The final complication is that we need to express that the two lifetimes are not independent: items borrowed from `self` must be valid for at least as long as the reference to `basename`. This means we end up with something like the following (it's possible there is a more concise way of expressing this, but I couldn't think of one):

```suggestion
    pub fn get_components_with_base<'b, 'a: 'b>(&'a self, basename: &'b str) -> impl Iterator<Item=&'a Component> + 'b {
        self.components.iter().filter(move |x| *x.base == *basename)
    }
```

chrissimpkins replied:
I started on this path but chickened out when I needed to add the `move` in order to make it work.  Does that move ownership out of `self.components` and can `self.components` be reused for access to the components data after  `get_components_with_base` execution?

cmyr replied:
I suspect someone else could explain this better, but basically the move only implicates the items referenced *inside* the closure that are defined outside the closure's scope, so in this case the `basename` argument. It's definitely confusing in this case, but I would take it to mean that the whole `&str` is moved into the closure, instead of being referenced (as an `&&str`?)

cmyr reviewed (approved):
Sorry for letting this sit, a few little suggestions but these seem like very reasonable additions :)

cmyr commented on src/glyph/tests.rs:
```diff
@@ -608,3 +608,98 @@ fn pointtype_from_str_trait() {
+}
+
+#[test]
+fn component_number() {
```
```suggestion
fn component_count() {
```

cmyr commented on src/glyph/tests.rs:
```diff
@@ -608,3 +608,98 @@ fn pointtype_from_str_trait() {
+fn component_number() {
+    let bytes = include_bytes!("../../testdata/MutatorSansLightWide.ufo/glyphs/A_dieresis.glif");
+    let glyph = parse_glyph(bytes).expect("initial load failed");
+    assert_eq!(glyph.component_number(), 2);
```
```suggestion
    assert_eq!(glyph.component_count(), 2);
```

cmyr commented on src/glyph/tests.rs:
```diff
@@ -608,3 +608,98 @@ fn pointtype_from_str_trait() {
+
+    let bytes = include_bytes!("../../testdata/MutatorSansLightWide.ufo/glyphs/A_.glif");
+    let glyph = parse_glyph(bytes).expect("initial load failed");
+    assert_eq!(glyph.component_number(), 0);
```
```suggestion
    assert_eq!(glyph.component_count(), 0);
```

cmyr reviewed (approved):
need to update the tests to reflect the name changes, I can do that from here I hope

chrissimpkins commented:
> need to update the tests to reflect the name changes, I can do that from here I hope

Thanks for this review Colin. No worries at all.  I'll update the tests to address the new method name and will update the `get_component_with_base` to return an iterator rather than a vector.  That will need test updates too.  I'll address it this week.

cmyr commented:
I updated the test names but didn't merge the iterator change. I'm going to do a release shortly, would you like this included?

chrissimpkins commented:
> I updated the test names but didn't merge the iterator change. I'm going to do a release shortly, would you like this included?

Feel free to release without these changes. I don't want to hold anything up and likely won't be able to address the tests until later this week.  I don't think that anything here is high priority.  Thanks Colin!

chrissimpkins commented:
All of Colin's changes were accepted, rebased on the recent changes in main (required f64 transition here), and updated tests. The CI should pass as of https://github.com/linebender/norad/pull/181/commits/d4493f7798181bbee5f4e4bc05dc30cbde2793c0.  If so, GTG from my standpoint.

cmyr reviewed (approved):
Thanks!

chrissimpkins commented:
Thanks for all of your help on this Colin!  Much appreciated.
